### PR TITLE
[Feature][MRV2] Support o_proj TP in model runner v2

### PR DESCRIPTION
**Pull request by xuchi-0808**
### What this PR does / why we need it?

Enables the `oproj_tensor_parallel_size` knob of fine-grained TP on model runner v2; it previously only worked on the V1 runner — on V2, graph-dispatched steps ran, but an eagerly dispatched step had no cross-DP alignment and would desynchronize the collectives.

o_proj TP shards the attention output projection along the **DP axis** (`vllm_ascend/distributed/parallel_state.py`). The exchange runs `all_to_all` → partial GEMM → `reduce_scatter` across that group on every attention forward, so every rank of the group must forward the **same number of tokens per step** — otherwise the generic exchange's HCCL collectives hang without an error.

With a graph mode active this invariant is native on V2: the upstream cudagraph dispatch (`sync_cudagraph_and_dp_padding` in `vllm/v1/worker/gpu/dp_utils.py`) agrees one group-wide token count and dispatches every rank to the same captured size. Idle DP ranks join through the engine-level dummy batch, which takes part in the same synchronization and runs the full forward — the o_proj exchange sits inside the forward, so unlike lmhead TP (#14668) no dummy collective call is needed. `_EXTRA_CTX.num_tokens` (consumed by the DSA/MLA o_proj exchange) is populated by the platform hook (`set_additional_forward_context`) with the same padded value V1 set explicitly. The exchange implementation, OTP group setup and static buffer sizing are runner-independent and carry over unchanged. The config validation is shared between the runners as well, and this PR tightens it rather than carrying it over (graph-mode criterion, PCP rejection — both under user-facing changes below).

The one gap is a step that degrades to **eager** — a mixed batch (eager only under `FULL_DECODE_ONLY`; under `FULL` it matches a mixed full graph and under `FULL_AND_PIECEWISE` a piecewise graph), a batch beyond the largest captured size (eager in every mode), or a request's boundary-recompute step on a PD decode node (the uniform-decode classifier treats it as prefill work, vLLM #51865, landed in v0.28.0). The eager branch of the DP sync keeps per-rank token counts, so the cross-DP collectives would hang. Rather than reject such steps, they are kept DP-aligned on the eager path too, with four small pieces and **no upstream binding replaced by this PR** (`vllm_ascend/worker/v2/eager_dp_padding.py` plus hooks in `vllm_ascend/worker/v2/model_runner.py`; the `dispatch_cg_and_sync_dp` rebinding further down that file is upstream PCP code, untouched here):

1. `execute_model` agrees the step's DP-group token max G with one CPU all_reduce before dispatch — real steps report their token count, dummy steps report zero. (V1's `_sync_metadata_across_dp` skips both the reduce and the padding wherever its MC2 path allows — dense models, or MoE with MC2 decode plus the recompute scheduler this knob requires — leaving per-step uniformity to the scheduler; when it cannot skip it still reduces and pads to the group max — with fine-grained TP on, which this knob implies, that pad condition always holds. The V2 eager branch pads nothing, so this hook agrees G explicitly.) Steps with zero scheduled tokens post nothing, matching the upstream early return;
2. `gather_batch_req_state` reports G as the batch's token count, so the dispatch input already carries the padded size and the eager branch hands every rank a descriptor at G *and* publishes `counts[dp_rank] == G`, keeping the forward-context invariant intact without touching the dispatch result;
3. `prepare_inputs` restores the real token extent for everything that treats `num_tokens` as real work (`query_start_loc` trailing fill, `InputBatch.num_tokens`); padded rows get `PAD_SLOT_ID` slot mappings (upstream `prepare_attn` → `compute_slot_mappings`, shared by the Ascend runner) and are never sampled (logits indices cover the real request rows only);
4. dummy steps rewrite their synthetic scheduler output to G, split into `decode_query_len`-sized requests so the dummy keeps the uniform-decode graph match when a capture size covers G — a single G-token request, a remainder request (non-uniform batch), or the over-`max_num_reqs` fallback (per-request counts at or above `decode_query_len`, non-uniform) each lose it, dragging the step to eager under `FULL_DECODE_ONLY` (the other modes still match their mixed/piecewise graphs). An idle rank's dummy batch then joins the collectives at the same shape.

The mechanism is installed only when o_proj TP runs with `dp_size > 1`; with the feature off, behavior is unchanged. That gate covers o_proj TP alone — mlp and embedding TP share the cross-DP constraint (same DP-axis exchange) and stay ungated here, to be covered by follow-up work with their own guards.

### Does this PR introduce _any_ user-facing change?

Yes. `oproj_tensor_parallel_size` now works on the V2 runner, functionally matching V1 for the supported PD deployment on the generic exchange (decode TPOT is currently ~2.7× worse than V1 at oproj=2; cause open, tracked under Performance): eager steps — mixed batches, oversized recompute chunks, request-arrival steps (i.e., the boundary-recompute step) on a PD decode node — are DP-padded and run instead of hanging the cross-DP collectives (mechanism UT-covered; the inflated case's e2e rerun is scheduled — see Testing). As before, the knob requires the PD decode-node deployment with `recompute_scheduler_enable=true`. As on V1, `enforce_eager` / `cudagraph_mode=NONE` are still rejected at config load; that precondition now tests `cudagraph_mode == NONE` instead of `model_config.enforce_eager`, so an explicit `cudagraph_mode=none` is caught as well (previously only `enforce_eager` was) — and since the validation is shared, the tightened criterion applies to the V1 runner too. An all-eager deployment stays out of scope — V1 rejects it for this knob as well; on the DSA path the static exchange buffers are sized at decode scale so oversized eager steps fail fast there, while the generic exchange has no capacity check of its own — an all-eager deployment of either exchange is rejected at config load. Spec decode combined with o_proj TP is unvalidated on either runner (see the testing notes). PCP (`prefill_context_parallel_size > 1`) is rejected alongside o_proj TP: its dispatch path recomputes the token count per rank, which would drop the DP padding and hang the cross-DP collectives on eager steps. That rejection is unconditional (fail rather than gamble on graph-only steps), and PCP is only reachable on the V2 runner on Ascend (the platform rejects it otherwise), so it is a new V2-specific limitation, not a parity regression.

### How was this patch tested?

Both e2e testbeds below exercise the generic `OProjRowParallelOp` exchange (MiniMax-M2.7, a GQA model, and the 4-layer dummy); the DSA path's static exchange is unchanged by this PR and runner-independent by construction, but unvalidated here (no e2e coverage).

- **UT** (CPU, pure mock): `tests/ut/worker/test_model_runner_v2_finegrained_tp.py` covers the DP-padding contract — the `execute_model` gate (feature on/off, dummy, profile, and the zero-token skip), group-max agreement, batch-state reporting at the group max with the real extent recorded, and the dummy-output rewrite (decode-shaped, exact total, capped at `max_num_reqs`); the `prepare_inputs` restore is pinned by a structural canary rather than behavioral coverage; `tests/ut/test_ascend_config.py` covers the config-level rejections (graph mode, PCP). The suite runs in CI cpu-ut, which builds against the main-verified vllm pin.
- **Single-A3 PD smoke** (decode node dp8×tp1, FULL_DECODE_ONLY + recompute scheduler, Mooncake connector, 4-layer dummy-weight model): continuous request arrival (KV transfer + boundary recompute) runs with the eager-padding hooks active — the eager branch executes without collective desync, and outputs are token-identical to the `oproj=0` baseline. The sampled steps carried at most one request per rank, so the gather-level inflation never fired (idle ranks' dummy batches do forward a row); that branch has UT coverage only, and no end-to-end run with real collectives exercises it yet.
- **2×A3 e2e** (MiniMax-M2.7-W8A8, PD tp8×dp2 / dp16×tp1, requests through the load-balance proxy): service up, sanity prompts normal, 640 requests at cc1–cc64 clean. That matrix predates the eager-padding mechanism — no eager step was observed across its 640 requests, so the padded path itself was never exercised and still needs a rerun.
- **Memory (net accounting)**: on the prior 2×A3 matrix (MiniMax-M2.7, tp8×dp2), enabling oproj=2 saved ~0.55 GiB of weights per rank and cost ~0.4–0.5 GiB of non-torch HCCL buffers for the OTP group (present on V1 as well; 0.48 GiB on that matrix, 0.42 GiB on a 4-layer dummy-weight single-A3 bed); oproj=4 measured ~0.82 GiB saved per rank on the same matrix, and larger splits were not measured. Net KV capacity is positive only for deep models (+256–384 KV tokens/rank at oproj=2 in that matrix, +1,536 at oproj=4; graph capture costs ~0.05 GiB); shallow models or small splits can be net negative.
- **Performance**: decode TPOT with the feature enabled is higher on V2 than on V1: on the single-A3 PD bed (dp8×tp1, n=8, engine-side) oproj=2 raised V2 from 6.17 to 16.76 ms (+10.6 ms, 2.7×) while V1 stayed flat (5.41→5.40 ms). Steady-state decode forwards one token per rank (no padding rows; each rank's log shows at least one num_tokens=1 full-graph replay), but the bed predates the dummy-rewrite fix and does not isolate the per-step CPU all_reduce the hooks add — whether the gap is in-graph OTP communication serialization, the extra per-step CPU reduce, or the prior-generation single-request dummy rewrite is open, tracked as follow-up work that also covers closing it.

The scenarios the runs above do not cover:

- the inflation itself (the group max above some rank's real count) under continuous request arrival in `FULL_DECODE_ONLY`, and with a deliberately undersized capture size so that ordinary steps overflow the largest captured graph — scheduled on the same test bed.
- spec decode: the drafter's main forward reuses the target's DP sync (`dp_sync` passthrough in `dispatch_cg_and_sync_dp`, no extra collective), which this padding aligns as well; its self-dispatched per-step decode runs its own sync. The combination is untested in this PR — there is no config guard for it either — and it is tracked as follow-up work.

- vLLM main: https://github.com/vllm-project/vllm/commit/b2f685834a6456197e7033966fdef52a23f1abcd

### CODE REVIEW

File: vllm_ascend/worker/v2/model_runner.py [hunk 1]
```diff
@@ -216,6 +251,16 @@ def execute_model(
             scheduler_output,
         )
 
+        validate_oproj_tp_graph_step(
+            scheduler_output.total_num_scheduled_tokens,
+            self.compilation_config.cudagraph_mode,
+            self.compilation_config.max_cudagraph_capture_size,
+            dp_size=self.dp_size,
```
[the remark this answers is not in the document]
Comment (xuchi-0808) [reply vllm_ascend/worker/v2/model_runner.py]: `self.dp_size` is assigned in the parent `GPUModelRunner.__init__`
(`vllm/v1/worker/gpu/model_runner.py:177` on the pinned v0.27.1:
`self.dp_size = self.parallel_config.data_parallel_size`), which
`NPUModelRunner.__init__` reaches via `super().__init__()`. Upstream itself
relies on this attribute on every step (`dispatch_cg_and_sync_dp(...,
self.dp_size, ...)` in `execute_model`), so it is always defined at runtime.
No change needed here.

File: tests/ut/worker/test_model_runner_v2_finegrained_tp.py [hunk 2]
```diff
@@ -236,3 +237,80 @@ def test_dummy_run_lmhead_disabled_or_profile_skips_collectives():
...
+def test_execute_model_wires_oproj_guard():
+    """The guard must run on every real (non-dummy, non-profile) step with the
+    runner's live dispatch inputs, so an eager fallback fails fast instead of
+    hanging the cross-DP collectives."""
+    runner = _make_runner()
+    runner.ascend_config = SimpleNamespace(
+        scheduler_config=SimpleNamespace(profiling_chunk_config=MagicMock())
+    )
+    runner.compilation_config = SimpleNamespace(cudagraph_mode=CUDAGraphMode.FULL, max_cudagraph_capture_size=512)
+    runner.dp_size = 2
```
[the remark this answers is not in the document]
Comment (xuchi-0808) [reply tests/ut/worker/test_model_runner_v2_finegrained_tp.py]: wrong review，not taking

**Comment by xuchi-0808** [comment]
/rerun
[Bot]: rerun completed.

**Rerun (failed jobs only):**
- E2E
